include/dali/dali.h

```diff
@@ -0,0 +1,698 @@
+ *                        if set, the dtype in the samples can be left at -1
+ * @param samples         the descriptors of the tensors to be attached to the TensorList;
+ *                        the `ndim` and `dtype` of the samples must match and they must match the
+ *                        values of `ndim` and `dtype` parameters.
+ * @param sample_deleters the deleters, one for each sample (or NULL)
+ *
+ * NOTE: If the sample_deleters specify the same object multiple times, its destructor must
+ *       internally use reference counting to avoid multiple deletion.
+ */
+DALI_API daliResult_t daliTensorListAttachSamples(
+  daliTensorList_h tensor_list,
+  int num_samples,
+  int ndim,
+  daliDataType_t dtype,
+  const daliTensorDesc_t *samples,
+  const daliDeleter_t *sample_deleters);
+
+
+/** Associates a stream with the TensorList.
+ *
+ * @param tensor_list   [in]  the TensorList whose buffer placement is queried
+ * @param out_placement [out] a pointer to a place where the return value is stored.
+ */
+DALI_API daliResult_t daliTensorListGetBufferPlacement(
+  daliTensorList_h tensor_list,
+  daliBufferPlacement_t *out_placement);
+
+/** Associates a stream with the TensorList.
+ *
+ * @param stream      an optional CUDA stream handle; if the handle poitner is NULL,
+ *                    host-synchronous behavior is prescribed.
+ * @param synchronize if true, the new stream (or host, if NULL), will be synchronized with the
+ *                    currently associated stream
+ */
+DALI_API daliResult_t daliTensorListSetStream(
+  daliTensorList_h tensor_list,
+  const cudaStream_t *stream,
+  daliBool synchronize
+);
+
+/** Gets the stream associated with the TensorList.
+ *
+ * @retval DALI_SUCCESS if the stream handle was stored in *out_stream
+ * @retval DALI_NO_DATA if the tensor list is not associated with any stream
+ *         error code otherwise
+ */
+DALI_API daliResult_t daliTensorListGetStream(
+  daliTensorList_h tensor_list,
+  cudaStream_t *out_stream
+);
+
+/** Gets the readiness event associated with the TensorList.
+ *
+ * @param tensor_list [in]  the tenosr list whose ready event is to be obtained
+ * @param out_event   [out] the pointer to the return value
+ *
+ * @retval DALI_SUCCESS if the ready event handle was stored in *out_event
+ * @retval DALI_NO_DATA if the tensor list is not associated with a readiness event
+ *         error code otherwise
+ */
+DALI_API daliResult_t daliTensorListGetReadyEvent(
+  daliTensorList_h tensor_list,
+  cudaEvent_t *out_event);
+
+/** Gets the readiness event associated with the TensorList or creates a new one.
+ *
+ * @param tensor_list   [in]  the tensor list to associate an even twith
+ * @param out_event     [out] optional, the event handle
+ *
+ * The function ensures that a readiness event is associated with the tensor list.
+ * It can also get the event handle, if the output parameter pointer is not NULL.
+ */
+DALI_API daliResult_t daliTensorListGetOrCreateReadyEvent(
```

Comment:
What are possible retvals and what do they mean are they the same as in the previous call?

Reply:
Well, at this point it's still a draft so I prefer not to commit to very precise error codes. When not specified otherwise, one can expect DALI_SUCCESS on success and some error code on failure. DALI_NO_DATA is specifically documented in some of the calls to denote more than one successful result.

Before we can commit to returning specific error codes. we need to fix error handling across DALI - which means nuking most of the DALI_ENFORCE/DALI_FAIL and replacing those with proper excetpions. Currently it's a mess - we mix usage errors (invalid argument or combination thereof), data errors (data-dependent index out of range, reduction over 0 extent), system errors (file not found, network errors), logic errors and more.